packages/block-editor/src/components/colors/use-colors.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * External dependencies
+ */
+import memoize from 'memize';
+import { startCase } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { Children, cloneElement, useMemo } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import InspectorControls from '../inspector-controls';
+import PanelColorSettings from '../panel-color-settings';
+import { useBlockEditContext } from '../block-edit';
+
+const createComponent = memoize( ( attribute, color ) => ( { children } ) =>
+	// Clone children, setting the style attribute from the color configuration,
+	// if not already set explicitly through props.
+	Children.map( children, ( child ) =>
+		cloneElement( child, {
+			style: { [ attribute ]: color, ...child.props.style },
+		} )
+	)
+);
+
+const createSetColor = memoize( ( setAttributes, name ) => ( newColor ) =>
+	setAttributes( { [ name ]: newColor } )
```

Comment:
We need to verify if we should set the named color attribute or the customColor attribute. Currently, there is a regression if I set the text color to a custom color I get the following markup:
```
<!-- wp:heading {"textColor":"#175d7d"} -->
<h2 class="has-175-d-7-d-color">fdgdfgdfgdf</h2>
<!-- /wp:heading -->
```

It is not possible to use a single attribute e.g: red is a valid custom color string but may also be a named color with a slightly different "red".

Reply:
Why do we need a differentiation between custom and not custom colors? It just makes validation fail and makes the usage of the colors API more complicated.

Reply:
I may set a custom color of "red" (is a valid CSS color), but also may have a named color "red" (or a previous theme may have had it in the past), how can I know if should set an inline style or use a class?

Reply:
We should stick to inline styles for this. Otherwise changing themes/color settings invalidates blocks.

Reply:
The reason the named colors were introduce was to avoid inline styles. We want to avoid styles as much as possible. Currently, changing themes/colors do not invalidate blocks, the named colors may be applied if the new theme does not define classes for them but that was something we wanted as the colors set when an old theme was used may not make sense in the new theme.

Reply:
See #16489

Reply:
The Pullquote problem is an exception because even for some named colors we want to use inline styles (for border styles which theme may not have defined as class) and depending on the block style the same color may be a border style (inline) or a background style (class), so the block works in different ways, the problem does not happen in other blocks.

Reply:
I guess the correct way to implement that functionality on pullquote would have been to have different attributes one for border and other for the background and depending on the block style we should options to pick one color or the other.
The problem with this other approach would have been that when changing styles the border-color would not be a background color right away and the previews also would not reflect the chosen colors.
We should think about ways to deal with the problem but it is a special situation and I'm not sure the API's we offer need to take that case in consideration.

Reply:
Yeah, but why not prevent developers from shooting themselves in the foot if we can?

I think that the fact that this happens is a code smell.

Reply:
The use of classes provides interesting use cases.
E.g.: if my site uses two colors primary and secondary right now I can change all the primary colors and secondary colors around my website including existing blocks by changing the color value in the class and in the palette.
How can we do something similar now and globally change colors on the websites?
People build interesting mechanisms, e.g., allow users to globally change block colors in all the blocks of the website from the customizer
https://richtabor.com/gutenberg-customizer-colors/.
Can we keep back-compatibility with these solutions when we use this hook in the paragraph and button blocks?

Reply:
That behavior belongs in a different API. Certain things have to be done differently to avoid invalidations when you use classes and I'd rather not overcomplicate a single hook when we can have a separate `useThemeColor` or something like that.